### ISSUE TITLE
Is dialtion_rate useful in keras DepthwiseConv2D layer ?

### ISSUE THREAD

**Issue by Pofatoezil** [bonlime/keras-deeplab-v3-plus#76]
Hi, I read the [document](https://www.tensorflow.org/api_docs/python/tf/keras/layers/DepthwiseConv2D)  and found dialtion_rate seems not a hyperparameter in DepthwiseConv2D layer.
But you used it in your [model](https://github.com/bonlime/keras-deeplab-v3-plus/blob/34e1a1dffd53a90c4de7d600ced3eb746c3dae1a/model.py#L79)
`x = DepthwiseConv2D((kernel_size, kernel_size), strides=(stride, stride), dilation_rate=(rate, rate)` 

I made a toy example to check this:
```
model=Sequential()
model.add(DepthwiseConv2D(3,strides=1,padding='valid',depth_multiplier=2,\
                          dilation_rate=(2,2),input_shape=(9,9,3))) 
model.summary()
```
Output:
```
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
depthwise_conv2d_6 (Depthwis (None, 7, 7, 6)           60        
=================================================================
Total params: 60
Trainable params: 60
Non-trainable params: 0
_________________________________________________________________
```
If dilation_rate is useful in DepthwiseConv2D , I think Output shape should be (5,5,6) not (7,7,6)

My English is pretty limited, plz don't mind

**Comment by bonlime** [comment]
The dilation rate IS useful in this model. The correct output shape in your example is (7,7,6), so it's working perfectly fine. The `dilation_rate` is not specified explicitly, but it's passed upstream to tf depthwise conv, you can understand it looking and the code of `DepthwiseConv2D` layer.

**Comment by Pofatoezil** [comment]
Thanks for your answer , I'll check the source code later.
But there are still something I don't understand.
In example, I don't use padding in this layer. The input shape is still (9,9,3).
I use kernel_size=3 ,strides=1 and dilation_rate=2 in this layer. So ,the effective kernel size should be 5
Why the output shape is (7,7,6)?

I test the dilation_rate in Conv2D layer
dilation rate =1
```
model_input=Input(shape=(9,9,3))
x=Conv2D(filters=3,kernel_size=3,strides=1,\
         dilation_rate=1,padding='valid')(model_input)
model=Model(model_input,x)
model.summary()

Output:
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_9 (InputLayer)         (None, 9, 9, 3)           0         
_________________________________________________________________
conv2d_8 (Conv2D)            (None, 7, 7, 3)           84        
=================================================================
Total params: 84
Trainable params: 84
Non-trainable params: 0
_________________________________________________________________
```
dilation rate =2
```
model_input=Input(shape=(9,9,3))
x=Conv2D(filters=3,kernel_size=3,strides=1,\
         dilation_rate=2,padding='valid')(model_input)
model=Model(model_input,x)
model.summary()

Output:
Layer (type)                 Output Shape              Param #   
=================================================================
input_10 (InputLayer)        (None, 9, 9, 3)           0         
_________________________________________________________________
conv2d_9 (Conv2D)            (None, 5, 5, 3)           84        
=================================================================
Total params: 84
Trainable params: 84
Non-trainable params: 0
________________________________________
```
If dilation_rate is work in DepthwiseConv2D layer, why output shape is (7,7,X) , not (5,5,X)?

**Comment by bonlime** [comment]
hmmm. maybe it automatically pads the input to preserve shape? You can check that output with `dilation_rate=2` is different from `dilation_rate=1`

**Comment by bonlime** [comment]
made a PR to main Keras branch https://github.com/keras-team/keras/pull/12526 which explicitly adds dilation rate to DepthWiseConv2d arguments

**Comment by Pofatoezil** [comment]
Hi, bonlime
After my test , the dilation rate is work in DepthwiseConv2D layers
But I found there is a bug about calculating output shape in Keras  when  I use the dilation_rate in DepthwiseConv2D.

In my test, I use Input shape = (9,9,channel) , kernel = 3, strides=1, dilation_rate=2
model.summary() shows the output shape is (7,7,channels), but the true output shape is (5,5,channels)

Thank you for discussing this problem with me!

**Comment by bonlime** [comment]
I'll test it tomorrow, but if there is a bug, you should open an issue in keras main repo or if you manage to fix - to make a pull request

**Comment by bonlime** [comment]
@Pofatoezil 
First of all, I was wrong in my first comment. The right shape in your example should be (5, 5, 6) as you said. I tested it on my machine and it works perfectly fine though. 

```
model=tf.keras.Sequential()
model.add(tf.keras.layers.DepthwiseConv2D(3,strides=1,padding='valid',depth_multiplier=2,\
                          dilation_rate=(2,2),input_shape=(9,9,3))) 
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
depthwise_conv2d (DepthwiseC (None, 5, 5, 6)           60        
=================================================================
Total params: 60
Trainable params: 60
Non-trainable params: 0

print("TF: {}. Keras: {}".format(tf.__version__, tf.keras.__version__))
>> TF: 2.0.0-dev20190317. Keras: 2.2.4-tf
```

**Comment by Pofatoezil** [comment]
My TF version is 1.9.0 , maybe the bug has been fixed in new version.
I'll upgrade my tensorflow , thank you.

**Comment by bonlime** [comment]
@Pofatoezil 
I found the source of bug you noticed and fixed it. The issue was that `DepthWiseConv2D().compute_shape` didn't support `dilation_rate`, and it led to the wrong `output_shape` calculation. Check the PR I mentioned above for additional info

**Comment by Pofatoezil** [comment]
Hi , @bonlime 
Thanks you for  check this bug and fixed it. 
In your testing example, the calculation of output shape is perfectly right .
The bug seems to be fixed in newest tenseorflow. 
But I am struggling in tensorflow-gpu vr 2.0.0a0 installation.
The problem is like [tensorflow#22794](https://github.com/tensorflow/tensorflow/issues/22794)
Would you mind telling me your tensorflow environment? 
Thank you

My environment:
OS Platform and Distribution: Windows 10 Pro
Mobile device: None
TensorFlow installed from: pip install
TensorFlow version: 2.0.0a0
Python Version: 3.6.6
Bazel version: not installed
CUDA/cuDNN version: CUDA 9.0, cuDNN 7.3
GPU model and memory: GF-GTX1060 6G
Visual Studio :visual studio 2015 community
